rust/gui-client/docs/process_split.md

```diff
@@ -1,34 +1,86 @@
 # Process split
 
-This is meant for Linux, but it will probably be similar on Windows.
+This is meant for Linux, but it will be similar on Windows.
+It is probably similar to the existing Mac / iOS / Android process split.
 
-## TODO
+## Split boundary
 
-- Are splitting the process for just the GUI client, or for CLI too?
+### Tunnel process
+
+Anything that requires CAP_NET_ADMIN or root privilege must happen inside the privileged tunnel process:
+
+- DNS control
+- Creating the tunnel interface
+- Adding / removing IP routes
+
+Some things don't need privileges, but it's more convenient to run them on the tunnel side to keep the IPC interface narrow:
+
+- DNS stub resolver
+- `get_system_default_resolvers`
+- Listening for DNS / network changes
+
+Anything that can work with normal user privileges should run in the non-privileged GUI process
+
+### GUI process
+
+- All GUI code, since GUIs in Linux generally don't work when running as root
+- Storing tokens in the secure keyring, since the desktop environment should be able to encrypt these with the user's password, which is more secure than keeping them in a file or securing them by TPM alone.
 
 ## Decisions
 
+- The CLI Client will not be split for now, it keep the same CLI interface but may be refactored. `firezone-client-tunnel` may replace it.
 - The CLI and GUI client packages will have a `conflicts` tag, at least for now.
 - Eventually, Clients and Gateways will be allowed to run on the same host, so avoid conflicts there
-- If the service account token exists, we are in service mode and won't accept connections from a GUI client, even if the token is invalid.
+- Binary names may change in the future.
+- The GUI stores tokens from interactive auth in a secure keyring provided by the desktop environment. (e.g. gnome-keyring) I assume the tunnel running as root with no desktop env, doesn't have access to that kind of security, since it starts without a password.
+- The GUI and privileged tunnel will live in separate binaries in case we need to set capabilities on the tunnel binary. This will also reduce the amount of code mapped into executable address space of the privileged tunnel process.
+
+## Files
+
+1. `$HOME/.config/autostart` - A link that auto-starts the GUI when the user logs in, created by `firezone-gui --register-auto-start`
+1. `/etc/dev.firezone.client/token` - A service account token, owned by root, with permissions 600. (rw- --- ---) Written by a human admin or an MDM on their behalf.
+1. `/usr/bin/firezone-client-tunnel` - A daemon that runs the tunnel, listens for commands from the GUI over IPC, and has elevated privilege. This is a systemd service on Linux and a Windows service on Windows. It takes over the role of the current Linux CLI Client.
```

Comment:
This may just be semantics but I would see it as: We "upgrade" the `firezone-linux-client` binary with the IPC functionality instead of the other way round and "downgrade" the tauri package to be just a GUI and depend on a running daemon in the background instead of splitting out another binary.

I think this may also be easier to ship because you can make incremental PRs that change the `firezone-linux-client` to what you need and then just delete a bunch of code from the tauri client.